File: leetcode-go/20.Valid-Parentheses/valid_parentheses.go
package main

/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
func isValid(s string) bool {
	if len(s)%2 == 1 {
		return false
	}
	m := map[byte]byte{
		')': '(',
		']': '[',
		'}': '{',
	}
	stack := []byte{}
	for i := 0; i < len(s); i++ {
		if _, ok := m[s[i]]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != m[s[i]] {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}
	return len(stack) == 0
}

// @lc code=end

func isValid1(s string) bool {
	if len(s)%2 == 1 {
		return false
	}
	stack := []byte{}
	for i := 0; i < len(s); i++ {
		if (s[i] == '(') || (s[i] == '[') || (s[i] == '{') {
			stack = append(stack, s[i])
		} else if ((s[i] == ')') && len(stack) > 0 && stack[len(stack)-1] == '(') ||
			((s[i] == ']') && len(stack) > 0 && stack[len(stack)-1] == '[') ||
			(s[i] == '}' && len(stack) > 0 && stack[len(stack)-1] == '{') {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}
	return len(stack) == 0
}

File: leetcode-go/020.Valid-Parentheses/valid_parentheses.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stack = [];
  let map = {
    ")": "(",
    "]": "[",
    "}": "{",
  };
  for (const c of s) {
    if (map[c]) {
      if (stack.length === 0 || stack[stack.length - 1] !== map[c]) {
        return false;
      }
      stack.pop();
    } else {
      stack.push(c);
    }
  }
  return stack.length === 0;
};
// @lc code=end

var isValid1 = function (s) {
  let stack = [];
  for (const c of s) {
    if (c === "(" || c === "[" || c === "{") {
      [];
      stack.push(c);
    } else if (
      (c === ")" && stack.length > 0 && stack[stack.length - 1] === "(") ||
      (c === "]" && stack.length > 0 && stack[stack.length - 1] === "[") ||
      (c === "}" && stack.length > 0 && stack[stack.length - 1] === "{")
    ) {
      stack.pop();
    } else {
      return false;
    }
  }
  return stack.length === 0;
};
